### PR TITLE
Use TWeakObjectPtr for unresolved objects

### PR DESCRIPTION
**Pull request by improbable-valentyn**
**Contributions**: We are not currently taking public contributions - see our [contributions](CONTRIBUTING.md) policy. However, we are accepting issues and we do want your [feedback](../README.md#give-us-feedback).

-------

#### Description
We queue up unresolved objects using just their raw object pointer. This is dangerous because the object could get deleted before it's resolved, which is potentially what causes issues like GCS-903 or GCS-904.
We should also check if we can remove the queued object entries because otherwise we're pretty much leaking that memory.
#### Primary reviewers
@m-samiec @Vatyx

### CODE REVIEW

File: SpatialGDK/Source/Private/Interop/SpatialSender.cpp [hunk 1]
```diff
@@ -597,16 +597,16 @@ Worker_ComponentUpdate USpatialSender::CreateMulticastUpdate(UObject* TargetObje
...
 
 	TSharedPtr<FRepLayout> RepLayout = NetDriver->GetFunctionRepLayout(Function);
 	RepLayout_SendPropertiesForRPC(*RepLayout, PayloadWriter, Parameters);
 
-	for (const UObject* Object : UnresolvedObjects)
+	for (TWeakObjectPtr<const UObject> Object : UnresolvedObjects)
 	{
 		// Take the first unresolved object
-		OutUnresolvedObject = Object;
+		OutUnresolvedObject = Object.Get();
```
Comment (joshuahuburn) [inline SpatialGDK/Source/Private/Interop/SpatialSender.cpp]: Perhaps we should change OutUnresolvedObject to also be a TWeakObjectPtr so it's usages are also protected somewhat.

File: SpatialGDK/Source/Private/Interop/SpatialSender.cpp [hunk 2]
```diff
@@ -559,18 +577,27 @@ Worker_CommandRequest USpatialSender::CreateRPCCommandRequest(UObject* TargetObj
...
+			// Take the first unresolved object
+			OutUnresolvedObject = Object.Get();
+			Schema_DestroyCommandRequest(CommandRequest.schema_type);
+			return CommandRequest;
+		}
+		else
+		{
+			// If the object is no longer valid (may have been deleted or IsPendingKill) then remove it from the UnresolvedObjects.
+			UnresolvedObjects.Remove(Object);
+			// TODO: Also remove it from other maps which reference the object by handle etc.
```
Comment (danielimprobable) [inline SpatialGDK/Source/Private/Interop/SpatialSender.cpp]: Jira ticket mention

File: SpatialGDK/Source/Private/Interop/SpatialSender.cpp [hunk 3]
```diff
@@ -524,14 +533,23 @@ void USpatialSender::QueueOutgoingUpdate(USpatialActorChannel* DependentChannel,
...
-		// Following up on the previous log: listing the unresolved objects
-		UE_LOG(LogSpatialSender, Log, TEXT("- %s"), *UnresolvedObject->GetName());
+			// Following up on the previous log: listing the unresolved objects
+			UE_LOG(LogSpatialSender, Log, TEXT("- %s"), *UnresolvedObject->GetName());
+		}
+		else
+		{
+			// If the object is no longer valid (may have been deleted or IsPendingKill) then remove it from the UnresolvedObjects.
+			Unresolved->Remove(UnresolvedObject);
+			// TODO: Also remove it from other maps which reference the object by handle etc.
```
Comment (danielimprobable) [inline SpatialGDK/Source/Private/Interop/SpatialSender.cpp]: Add jira reference to comment

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp [hunk 4]
```diff
@@ -482,20 +482,29 @@ void USpatialSender::ResetOutgoingUpdate(USpatialActorChannel* DependentChannel,
...
+				{
+					ObjectToUnresolved.Remove(UnresolvedObject);
+				}
 			}
 		}
+		else
+		{
+			// If the object is no longer valid (may have been deleted or IsPendingKill) then remove it from the UnresolvedObjects.
+			Unresolved->Remove(UnresolvedObject);
+			// TODO: Also remove it from other maps which reference the object by handle etc.
```
Comment (girayimprobable) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp]: Mention ticket number

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp [hunk 5]
```diff
@@ -601,18 +628,27 @@ Worker_ComponentUpdate USpatialSender::CreateMulticastUpdate(UObject* TargetObje
...
+			// Take the first unresolved object
+			OutUnresolvedObject = Object.Get();
+			Schema_DestroyComponentUpdate(ComponentUpdate.schema_type);
+			return ComponentUpdate;
+		}
+		else
+		{
+			// If the object is no longer valid (may have been deleted or IsPendingKill) then remove it from the UnresolvedObjects.
+			UnresolvedObjects.Remove(Object);
+			// TODO: Also remove it from other maps which reference the object by handle etc.
```
Comment (danielimprobable) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp]: Mention ticket number here as well

**Pull request merged**